flink-streaming-java/src/test/java/org/apache/flink/streaming/api/functions/source/FileSourceFunctionTest.java

```diff
@@ -0,0 +1,258 @@
+	public void testNormalOp() throws Exception {
+		testForEmptyLocation(false);
+	}
+
+	@Test
+	public void testCancelation() throws Exception {
+		testForEmptyLocation(true);
+	}
+
+	private void testForEmptyLocation(final boolean midCancel) throws Exception {
+
+		final int noOfSplits = 5;
+		final int cancelAt = 2;
+
+		final LifeCycleTestInputFormat format = new LifeCycleTestInputFormat();
+		final FileSourceFunction<Integer> reader = new FileSourceFunction<>(format, TypeInformation.of(Integer.class));
+		reader.setRuntimeContext(new MockRuntimeContext(format, noOfSplits));
+
+		Assert.assertTrue(!format.isConfigured);
+		Assert.assertTrue(!format.isInputFormatOpen);
+		Assert.assertTrue(!format.isSplitOpen);
+
+		reader.open(new Configuration());
+		Assert.assertTrue(format.isConfigured);
+
+		TestSourceContext ctx = new TestSourceContext(reader, format, midCancel, cancelAt);
+		reader.run(ctx);
+
+		int splitsSeen = ctx.getSplitsSeen();
+		Assert.assertTrue(midCancel ? splitsSeen == cancelAt : splitsSeen == noOfSplits);
+
+		// we have exhausted the splits so the
+		// format and splits should be closed by now
+
+		Assert.assertTrue(!format.isSplitOpen);
+		Assert.assertTrue(!format.isInputFormatOpen);
+	}
+
+
+	private static class LifeCycleTestInputFormat extends RichInputFormat<Integer,InputSplit> {
+
+		private boolean isConfigured = false;
+		private boolean isInputFormatOpen = false;
+		private boolean isSplitOpen = false;
+
+		// end of split
+		private boolean eos = false;
+
+		private int splitCounter = 0;
+
+		@Override
+		public void openInputFormat() {
+			this.isInputFormatOpen = true;
+		}
+
+		@Override
+		public void closeInputFormat() {
+			this.isInputFormatOpen = false;
+		}
+
+		@Override
+		public void configure(Configuration parameters) {
+			this.isConfigured = true;
+		}
+
+		@Override
+		public BaseStatistics getStatistics(BaseStatistics cachedStatistics) throws IOException {
+			return null;
+		}
+
+		@Override
+		public InputSplit[] createInputSplits(int minNumSplits) throws IOException {
+			InputSplit[] splits = new InputSplit[minNumSplits];
+
+			for (int i = 0; i < minNumSplits; i++) {
+				final int idx = i;
+				splits[idx] = new InputSplit() {
+					@Override
+					public int getSplitNumber() {
+						return idx;
+					}
+				};
+			}
+			return splits;
+		}
+
+		@Override
+		public InputSplitAssigner getInputSplitAssigner(InputSplit[] inputSplits) {
+			return null;
+		}
+
+		@Override
+		public void open(InputSplit split) throws IOException {
+			// whenever a new split opens,
+			// the previous should have been closed
+			Assert.assertTrue(!isSplitOpen);
+
+			isSplitOpen = true;
+			eos = false;
+		}
+
+		@Override
+		public boolean reachedEnd() throws IOException {
+			return eos;
+		}
+
+		@Override
+		public Integer nextRecord(Integer reuse) throws IOException {
+			eos = true;
+			return splitCounter++;
+		}
+
+		@Override
+		public void close() throws IOException {
+			this.isSplitOpen = false;
```

Comment:
The problem with this is that the close() in the FileSourceFunction is called also in the finally{} block, which can be redundant in the cases that the split was already closed. This is why I do not
check it here, but I check that whenever you call open(), the format was not opened before.